Update CI scripts to lint early and fail fast

jasonrudolph opened the pull request:
### Issue or RFC Endorsed by Atom's Maintainers

This pull request closes #18940.

This pull request also helps us make progress toward #19006.

### Description of the Change

This pull request enhances the CI builds to "fail fast" when a linting violation is present, so that developers can enjoy a faster feedback loop.  (See #18940 for more discussion of the motivation.)

As @50Wliu mentions in https://github.com/atom/atom/issues/18940#issuecomment-475021141, `script/lint` depends on `script/bootstrap`. Prior to this pull request, `script/bootstrap` only ran as part of `script/build`. Because of that, we ran `script/lint` *after* `script/build` so that all of the dependencies for `script/lint` would be in place. However, `script/build` is one of the slowest parts of the CI build, so it would be nice if we didn't have to wait for `script/build` to complete in order to check for linting violations. 

With the changes in this pull request, we now run `script/bootstrap` on its own early in the CI set-up, and that enables us to run `script/lint` _before_ running `script/build`. 😅

### Alternate Designs

Since `script/bootstrap` now runs earlier in the CI process, this pull request adds an option that allows you to tell `script/build` to skip bootstrapping. This isn't essential, but I think it's nice to have. As an alternative, instead of adding an option to `script/build` to allow you to skip bootstrapping, we could have `script/build` always bootstrap (just as it did prior to this pull request). This approach would result in less logic to maintain, but the unnecessary bootstrapping adds about 15 seconds to the build (as measured on a 2018 MacBook Pro). Since we're running `script/lint` before `script/build`, and since `script/lint` depends on `script/bootstrap`, we don't _need_ to run `script/bootstrap` as part of `script/build`, so we can save some time by skipping it.

### Possible Drawbacks

It introduces a small bit of additional conditional logic to `script/build`.

### Verification Process

To verify these changes, I [intentionally inserted a linting violation](https://github.com/atom/atom/commit/2717815762823d2d9bc75d2a9e2712845739c587) and confirmed that the builds failed fast. I then [removed the linting violation](https://github.com/atom/atom/commit/b5748bcbdc07928b85735e10d46470e5c52cd17e) and confirmed that all the builds passed. For more details, please see below.

- [x] Travis CI
	- [x] [Fail fast on linting violation](https://travis-ci.org/atom/atom/builds/514681544#L736)
	- [x] [Complete successfully after removing linting violation](https://travis-ci.org/atom/atom/builds/514776458#L736)
- [x] Azure DevOps Production Branches Pipeline
	- [x] Linux
		- [x] [Fail fast on linting violation](https://github.visualstudio.com/Atom/_build/results?buildId=37020&view=logs&jobId=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&taskId=5b4cc83a-7bb0-5664-5bb1-588f7e4dc05b&lineStart=14&lineEnd=15&colStart=1&colEnd=1)
		- [x] [Complete successfully after removing linting violation](https://github.visualstudio.com/Atom/_build/results?buildId=37090)
	- [x] Windows 64
		- [x] [Fail fast on linting violation](https://github.visualstudio.com/Atom/_build/results?buildId=37020&view=logs&jobId=4fbced83-508e-5fe0-c978-5c71ec0fc506&taskId=efea9dc6-8b7a-5cfd-995a-4727b0e8449d&lineStart=14&lineEnd=15&colStart=1&colEnd=1)
		- [x] [Complete successfully after removing linting violation](https://github.visualstudio.com/Atom/_build/results?buildId=37090)
	- [x] Windows x86
		- [x] [Fail fast on linting violation](https://github.visualstudio.com/Atom/_build/results?buildId=37020&view=logs&jobId=a5e52b91-c83f-5429-4a68-c246fc63a4f7&taskId=2dac33d1-b0ca-5feb-3c36-18b56319df50&lineStart=14&lineEnd=15&colStart=1&colEnd=1)
		- [x] [Complete successfully after removing linting violation](https://github.visualstudio.com/Atom/_build/results?buildId=37090)
	- [x] macOS
		- [x] [Fail fast on linting violation](https://github.visualstudio.com/Atom/_build/results?buildId=37020&view=logs&jobId=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&taskId=5b4cc83a-7bb0-5664-5bb1-588f7e4dc05b&lineStart=14&lineEnd=15&colStart=1&colEnd=1)
		- [x] [Complete successfully after removing linting violation](https://github.visualstudio.com/Atom/_build/results?buildId=37090)
- [x] Azure DevOps Pull Requests Pipeline
	- [x] Linux
		- [x] [Fail fast on linting violation](https://github.visualstudio.com/Atom/_build/results?buildId=37011&view=logs&jobId=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&taskId=5b4cc83a-7bb0-5664-5bb1-588f7e4dc05b&lineStart=14&lineEnd=15&colStart=1&colEnd=1)
		- [x] [Complete successfully after removing linting violation](https://github.visualstudio.com/Atom/_build/results?buildId=37090)
	- [x] Windows 64
		- [x] [Fail fast on linting violation](https://github.visualstudio.com/Atom/_build/results?buildId=37011&view=logs&jobId=4fbced83-508e-5fe0-c978-5c71ec0fc506&taskId=efea9dc6-8b7a-5cfd-995a-4727b0e8449d&lineStart=14&lineEnd=15&colStart=1&colEnd=1)
		- [x] [Complete successfully after removing linting violation](https://github.visualstudio.com/Atom/_build/results?buildId=37090)
	- [x] Windows x86
		- [x] [Fail fast on linting violation](https://github.visualstudio.com/Atom/_build/results?buildId=37011&view=logs&jobId=97a617bf-bcbd-5dfa-bba2-cfba2747b693&taskId=ce76a027-d404-5bfb-7945-43613713e966&lineStart=15&lineEnd=16&colStart=1&colEnd=1)
		- [x] [Complete successfully after removing linting violation](https://github.visualstudio.com/Atom/_build/results?buildId=37090)
	- [x] macOS
		- [x] [Fail fast on linting violation](https://github.visualstudio.com/Atom/_build/results?buildId=37011&view=logs&jobId=a5e52b91-c83f-5429-4a68-c246fc63a4f7&taskId=2dac33d1-b0ca-5feb-3c36-18b56319df50&lineStart=14&lineEnd=15&colStart=1&colEnd=1)
		- [x] [Complete successfully after removing linting violation](https://github.visualstudio.com/Atom/_build/results?buildId=37090)

Note: This pull request doesn't change anything about the AppVeyor build.  In the AppVeyor build, `script/lint` and `script/test` both run as part of the `test_script` step [here](https://github.com/atom/atom/blob/44271255fc5e0ff6ab73c49a669f0c068837947a/appveyor.yml#L68).  To have `script/lint` run before `script/build`, I think we'd have to move `script/build` out of the [`build_script` step](https://github.com/atom/atom/blob/44271255fc5e0ff6ab73c49a669f0c068837947a/appveyor.yml#L42-L64) into the `test_script` step, and that just seems weird. Given that we're hoping to retire AppVeyor before too long, I think we should leave this as is. Linting will still fail fast on Travis CI and Azure DevOps, so that will give developers the faster feedback loop that we're looking for.

### Release Notes

N/A - This is not a user-facing change, so it probably shouldn't go into the release notes.

---

/fyi @Aerijo

jasonrudolph committed:
Add back bootstrapping to script/build on Windows

Hopefully this will resolve the error we're seeing here:

https://github.visualstudio.com/Atom/_build/results?buildId=37021&view=logs&jobId=97a617bf-bcbd-5dfa-bba2-cfba2747b693&taskId=1e1369e9-1c56-59b0-830c-f1606a680c8a&lineStart=13&lineEnd=19&colStart=1&colEnd=1

🤞

jasonrudolph committed:
Lint earlier in the build process

jasonrudolph committed:
Insert linter violation (for testing purposes)

jasonrudolph committed:
Teach linting to fail fast on Travis CI

jasonrudolph committed:
Revert "Insert linter violation (for testing purposes)"

This reverts commit 2717815762823d2d9bc75d2a9e2712845739c587.

jasonrudolph commented on .travis.yml:
```diff
@@ -28,11 +28,12 @@ install:
-  - script/lint
-  - script/test
+script: >
+  script/lint &&
```
If linting fails, we want the build to halt immediately. In order to make that happen Travis CI, we need to combine all of the commands together. /xref https://github.com/travis-ci/travis-ci/issues/6188

50Wliu commented:
> the unnecessary bootstrapping adds about 15 seconds to the build

Not necessarily related to this PR, but I noticed that we don't seem to be using `npm ci` on any of our CI instances. Maybe that would eliminate this difference or make it negligible?

jasonrudolph committed:
Provide the BUILD_ARCH env var for bootstrapping Windows env on VSTS

`script/vsts/windows-run.js` checks the BUILD_ARCH env var, so let's 
make sure it's populated.

jasonrudolph commented:
> Not necessarily related to this PR, but I noticed that we don't seem to be using `npm ci` on any of our CI instances.

Good point, @50Wliu. I could definitely see some benefit in investigating that in a follow-up PR. :bow:
